### ISSUE TITLE
Protostuff reorder array with null values upon deser

### ISSUE THREAD

**Issue by koen-serneels** [protostuff/protostuff#266]
Hi. Protostuff seems to 'reorder' positions in an array, at least when 'null' values are involved:

```
          System.setProperty("protostuff.runtime.always_use_sun_reflection_factory", "true");
	  System.setProperty("protostuff.runtime.collection_schema_on_repeated_fields", "true")
	  System.setProperty("protostuff.runtime.use_sun_misc_unsafe", "false");

	   class Something {
			private Object[] o;
		}

		Something something = new Something();
		something.o = new Object[] { new Long(1), "testA", null, "testB" };

		Schema schema = RuntimeSchema.createFrom(Something.class);
		LinkedBuffer buffer = LinkedBuffer.allocate(512);
		ByteArrayOutputStream bos = new ByteArrayOutputStream();
		final byte[] protostuff;
		try {
			protostuff = ProtostuffIOUtil.toByteArray(something, schema, buffer);
			bos.write(protostuff);
		} finally {
			buffer.clear();
			bos.close();
		}
		schema = RuntimeSchema.getSchema(Something.class);
		Something result = (Something) schema.newMessage();
		ProtostuffIOUtil.mergeFrom(new ByteArrayInputStream(bos.toByteArray()), result, schema);
```
In the derserialized array the order is '1', 'testA', 'testB' , null. What are we missing? Is this a bug, a setting that is not there?

**Comment by dyu** [comment]
Set ```protostuff.runtime.allow_null_array_element``` to true

**Comment by koen-serneels** [comment]
Ok thanks, is there any other setting that we might be missing?

**Comment by koen-serneels** [comment]
I added the system property to the above code, but but the problem remains: the null value is shifted to the end of the array. Any ideas?

**Comment by dyu** [comment]
Set those system properties on application startup ... before the classloading of protostuff-runtime

**Comment by koen-serneels** [comment]
Did that, but no go:
`
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;

import io.protostuff.LinkedBuffer;
import io.protostuff.ProtostuffIOUtil;
import io.protostuff.Schema;
import io.protostuff.runtime.RuntimeSchema;

public class SerializationTest {

	public static void main(String[] args) throws Exception {

		class Something {
			private Object[] o;
		}

		Something something = new Something();
		something.o = new Object[] { new Long(1), "testA", null, "testB" };

		Schema schema = RuntimeSchema.createFrom(Something.class);
		LinkedBuffer buffer = LinkedBuffer.allocate(512);
		ByteArrayOutputStream bos = new ByteArrayOutputStream();
		final byte[] protostuff;
		try {
			protostuff = ProtostuffIOUtil.toByteArray(something, schema, buffer);
			bos.write(protostuff);
		} finally {
			buffer.clear();
			bos.close();
		}
		schema = RuntimeSchema.getSchema(Something.class);
		Something result = (Something) schema.newMessage();
		ProtostuffIOUtil.mergeFrom(new ByteArrayInputStream(bos.toByteArray()), result, schema);

		System.out.println("Should be: 1 - testA - null - TestB");
		System.out.println("Actual result: " + result.o[0] + " - " + result.o[1] + " - " + result.o[2] + " - " + result.o[3]);
	}
}

`
>  2019-04-17 23:15:29 ⌚  koen-HP-ZBook-15-G3 in /tmp
> ○ → java -version
> java version "1.8.0_151"
> Java(TM) SE Runtime Environment (build 1.8.0_151-b12)
> Java HotSpot(TM) 64-Bit Server VM (build 25.151-b12, mixed mode)
> 
>  2019-04-17 23:15:31 ⌚  koen-HP-ZBook-15-G3 in /tmp
> ○ → java -Dprotostuff.runtime.always_use_sun_reflection_factory=true -Dprotostuff.runtime.collection_schema_on_repeated_fields=true -Dprotostuff.runtime.use_sun_misc_unsafe=false -Dprotostuff.runtime.allow_null_array_element=true -cp ./:protostuff-uberjar-1.5.9.jar SerializationTest
> Should be: 1 - testA - null - TestB
> Actual result: 1 - testA - testB - null
> 
>  2019-04-17 23:15:45 ⌚  koen-HP-ZBook-15-G3 in /tmp
> ○ → 
>

**Comment by koen-serneels** [comment]
Any update on this?

**Comment by dyu** [comment]
You found a bug wherein the non-null array nature is not applied when the field is ```Object[]```.  Changing that to scalars (String[]) or pojos (Foo[]) would work.   I'll take a closer look. Sorry for the late reply

**Comment by dyu** [comment]
Fixed with commit: 8fddbd526968ace00583a3abfd64b9287351cea0
